cwr/parser.go

```diff
@@ -0,0 +1,90 @@
+// This file is part of the go-meta library.
+//
+// Copyright (C) 2017 JAAK MUSIC LTD
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+//
+// If you have any questions please contact yo@jaak.io
+
+package cwr
+
+import (
+	"encoding/json"
+	"errors"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"strings"
+)
+
+// ParseCWRFile parse a give cwr file and returns an array of registeredWorks
+func ParseCWRFile(cwrFilePath string) (registeredWorks []*RegisteredWork, err error) {
+
+	//phase 1 : Transform cwr formatted file to cwr-json file using CWR-DataApi python script.
+	//get the absolute path the cwrfile.
+	cwrFilePath, err = filepath.Abs(cwrFilePath)
+	if err != nil {
+		return nil, err
+	}
+	_, b, _, ok := runtime.Caller(0)
+	if !ok {
+		return nil, errors.New("error getting base path")
+	}
+	basePath := filepath.Dir(b)
+	cwr2jsonpy := basePath + "/CWR-DataApi/cwr2json.py"
+	outputFilePath := basePath + "/testdata/testfile.json"
+
+	cmd := exec.Command("python3", cwr2jsonpy, "-i", cwrFilePath, "-o", outputFilePath)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	log.Println(cmd.Run())
+
+	filedata, err := ioutil.ReadFile(outputFilePath)
+
+	if err != nil {
+		return nil, err
+	}
+	var data map[string]interface{}
+	if err := json.Unmarshal(filedata, &data); err != nil {
+		return nil, err
+	}
+
+	for _, group := range data["transmission"].(map[string]interface{})["groups"].([]interface{}) {
+		for _, tx := range group.(map[string]interface{})["transactions"].([]interface{}) {
+			var registeredWork *RegisteredWork
+
+			for _, record := range tx.([]interface{}) {
+
+				if record.(map[string]interface{})["record_type"] == "REV" ||
+					record.(map[string]interface{})["record_type"] == "NWR" {
```

Comment:
Rather than this deeply nested loop through `map[string]interface{}` and `[]interface{}`, just define a Go type with JSON tags to deserialize into, then you also get error checking for free from the JSON library.

Reply:
Anyhow it needs to iterate through groups, transactions and records.